src/main/java/cn/nukkit/entity/passive/EntityVillager.java

```diff
@@ -43,6 +57,134 @@ public String getName() {
+        this.recipes = new ArrayList<TradeInventoryRecipe>();
+        this.dataProperties.putLong(DATA_TRADING_PLAYER_EID, 0L);
+        if(this.namedTag.contains("Offers")) {
+            ListTag<CompoundTag> nbtRecipes = getListRecipes();
+            for(CompoundTag nbt : nbtRecipes.getAll()) {
+                recipes.add(TradeInventoryRecipe.toNBT(nbt));
+            }
+        } else {
+            CompoundTag nbt = new CompoundTag("Offers");
+            nbt.putList(new ListTag<CompoundTag>("Recipes"));
+            nbt.putList(getDefaultTierExpRequirements());
+            this.namedTag.putCompound("Offers", nbt);
+        }
+    }
+    
+    public void setTraderName(String name) {
+        this.namedTag.putString("TraderName", name);
+    }
+    
+    public String getTraderName() {
+        if(!this.namedTag.contains("TraderName")) {
+            this.namedTag.putString("TraderName", getNameTag());
+        }
+        return this.namedTag.getString("TraderName");
+    }
+    
+    public void setTradeTier(int tier) {
+    	if(tier > 4){
+    	    throw new ServerException("Maximal tier is 4, but your tier is higher than 4");
+    	}
+        this.namedTag.putInt("TradeTier", tier);
+        this.dataProperties.putInt(DATA_TRADE_TIER, tier);
+    }
+    
+    public void setMaxTradeTier(int maxTier) {
+    	if(maxTier > 4) maxTier = 4;
+        this.dataProperties.putInt(DATA_MAX_TRADE_TIER, maxTier);
+    }
+    
+    public int getMaxTradeTier() {
+        if(!this.dataProperties.exists(DATA_MAX_TRADE_TIER)) {
+            this.setMaxTradeTier(4);
+        }
+        return this.dataProperties.getInt(DATA_MAX_TRADE_TIER);
+    }
+    
+    public void setExperience(int experience) {
+        this.dataProperties.putInt(DATA_TRADE_EXPERIENCE, experience);
+    }
+    
+    public int getExperience() {
+        if(!this.dataProperties.exists(DATA_TRADE_EXPERIENCE)) {
+            setExperience(0);
+        }
+        return this.dataProperties.getInt(DATA_TRADE_EXPERIENCE);
+    }
+    
+    public int getTradeTier() {
+        if(!this.namedTag.contains("TradeTier")) {
+            this.namedTag.putInt("TradeTier", 0);
+        }
+        return this.namedTag.getInt("TradeTier");
+    }
+    
+    public void setWilling(boolean value) {
+        this.namedTag.putBoolean("Willing", value);
+    }
+    
+    public boolean isWilling() {
+        if(!this.namedTag.contains("Willing")) {
+            this.namedTag.putBoolean("Willing", true);
+        }
+        return this.namedTag.getBoolean("Willing");
+    }
+    
+    public void cancelTradingWithPlayer() {
+    	this.setTradingWith(0L);
+    }
+    
+    public void setTradingWith(long eid) {
+        this.dataProperties.putLong(DATA_TRADING_PLAYER_EID, eid);
+    }
+    
+    public boolean isTrading() {
+        return this.dataProperties.getLong(DATA_TRADING_PLAYER_EID) != 0L;
+    }
+    
+    @Override
+    public boolean onInteract(Player player, Item item) {
```

Comment:
(Google Translate)
I think we also need to rewrite the onInteract(Player player, Item item, Vector3 clickedPos) method
like this:
```
    @Override
    public boolean onInteract(Player player, Item item, Vector3 clickedPos) {
        if (item.getId() == Item.NAME_TAG) {
            return super.onInteract(player, item, clickedPos);
        }
        return this.onInteract(player, item);
    }
```